Added topic on libs working together

PeterTurcan opened the pull request.

alandefreitas commented on user-guide/modules/ROOT/nav.adoc:
```diff
@@ -1,5 +1,6 @@
 * xref:intro.adoc[Introduction to Boost]
 * xref:getting-started.adoc[]
+* xref:libraries-that-work-together.adoc[]
```
I find it weird to take room in the level 0 hierarchy for such a specific type of content, especially since the intro.adoc still has a very poor and busy visual structure. I would imagine this kind of thing in an admonition in intro.adoc or something rather than as a level 0 section.

alandefreitas replied:
@anarthal feedback?

PeterTurcan replied:
Could move it to the intro.

vinniefalco replied:
We need to really dial down the length of these nav items.

anarthal replied:
@alandefreitas agree. I'd expect a "this is what you can find in Boost" section instead. I've never seen such a section before anywhere in other libraries.

alandefreitas commented on user-guide/modules/ROOT/pages/libraries-that-work-together.adoc:
```diff
@@ -0,0 +1,32 @@
+
+Boost libraries each has its own use, but many are designed to work together to simplify complex tasks. Here are a few examples of Boost libraries that work well together:
+
+== Working Well in Pairs
```
boost.json and boost.describe seem like a trendy combination lately.

PeterTurcan replied:
OK - I will add that as a pair.

PeterTurcan replied:
And is the combined purpose to use Boost.Describe to automatically generate code to serialize and deserialize custom data types to and from JSON? Or is there another reason for the combo?

alandefreitas replied:
Yes. I think that's it.

anarthal replied:
Describe is meant to add reflection to stuff. A useful consequence of it is being able to serialize or deserialize to JSON. But you can use it to add operators like `operator==` and `operator<<` (really useful in unit tests). There is a (almost finished) Boost.MySQL feature that uses Describe as an interface for custom row types. So the end user creates a struct annotated with Boost.Describe macros, and Boost.MySQL is automatically able to parse the results of a `SELECT` query into that type. You can use it to convert a struct into a tuple, too. In general, anything using `std::tuple` could instead use a Boost.Describe struct.

PeterTurcan replied:
@anarthal Good info, we can add the SQL example when the lib is released.

alandefreitas replied:
Boost.MySQL and Describe is a good idea too. I didn't know about this feature. That's great. 

I think Boost.Json and Describe came to mind because I remember them saying they were working on that. But I don't remember if that's a feature in Boost.Describe or Boost.Json.

alandefreitas commented on user-guide/modules/ROOT/pages/libraries-that-work-together.adoc:
```diff
@@ -0,0 +1,32 @@
+== Working Well in Multiple Combinations
+
+[disc]
+* boost:serialization[] and boost:filesystem[], boost:multi_array[], or or boost:ptr_container[] : Boost.Serialization is used for saving and loading data, which is a common task in many types of software, including games. After you've serialized your data with Boost.Serialization, you'll often want to write it to disk. Boost.Filesystem provides a portable and easy-to-use interface for file and d ...
```
> Boost.Serialization. Boost.Pointer_Container

No links

PeterTurcan replied:
Linked on first use:  boost:serialization[] and boost:filesystem[], boost:multi_array[], or or boost:ptr_container[] ?

Could make all references links.

alandefreitas replied:
Yes. The other names without links look bad.

alandefreitas commented on user-guide/modules/ROOT/pages/libraries-that-work-together.adoc:
```diff
@@ -0,0 +1,32 @@
+[disc]
+* boost:serialization[] and boost:filesystem[], boost:multi_array[], or or boost:ptr_container[] : Boost.Serialization is used for saving and loading data, which is a common task in many types of software, including games. After you've serialized your data with Boost.Serialization, you'll often want to write it to disk. Boost.Filesystem provides a portable and easy-to-use interface for file and d ...
+
+* boost:json[] and boost:optional[] or boost:variant[] : Boost.Json is a library for parsing, generating, and manipulating JSON data. Boost.Optional can be used to represent optional JSON fields. JSON objects often have optional fields, and Boost.Optional can be used to represent these in a type-safe way. Boost.Variant is a safe, generic, stack-based discriminated union container, which can repre ...
```
Same

alandefreitas commented on user-guide/modules/ROOT/pages/libraries-that-work-together.adoc:
```diff
@@ -0,0 +1,32 @@
+
+* boost:json[] and boost:optional[] or boost:variant[] : Boost.Json is a library for parsing, generating, and manipulating JSON data. Boost.Optional can be used to represent optional JSON fields. JSON objects often have optional fields, and Boost.Optional can be used to represent these in a type-safe way. Boost.Variant is a safe, generic, stack-based discriminated union container, which can repre ...
+
+* boost:smart_ptr[] and boost:thread[], boost:signals2[], boost:flyweight[], boost:interprocess[] : Boost.Smart_Ptr is a library that provides classes for smart pointer objects, which are objects that behave like regular pointers but also have the property that they automatically delete the object they manage (if any) at the appropriate time. Smart pointers are a key component of C++ resource man ...
```
Same... actually that happens in all paragraphs.

PeterTurcan commented:
@alandefreitas I have updated the PR based on the feedback.

alandefreitas commented on user-guide/modules/ROOT/pages/boost-history.adoc:
```diff
@@ -48,9 +48,7 @@ These libraries, among others, showcase the value and versatility of Boost in pr
-The Boost library is managed by the https://cppalliance.org/[pass:[C++] Alliance], with a mission "to make the C++ programming language accessible and useful to anyone who wishes to learn and apply the language."
-
-The Boost pass:[C++] Library has come a long way since its inception in the late 1990s, to a widely used and influential https://www.boost.org/doc/libs/[collection of over 180 libraries]. Boost has played a crucial role in shaping the pass:[C++] language and its ecosystem. As the pass:[C++] language continues to evolve, the Boost pass:[C++] Library will undoubtedly remain an essential and innovat ...
+Boost pass:[C++] has come a long way since its inception in the late 1990s, to a widely used and influential https://www.boost.org/doc/libs/[collection of over 180 libraries]. Boost has played a crucial role in shaping the pass:[C++] language and its ecosystem. As the pass:[C++] language continues to evolve, the Boost pass:[C++] Library will undoubtedly remain an essential and innovative part of  ...
```
> Boost pass:[C++]

I never heard the term "Boost C++". It's not what it's called.

It's either "The Boost Libraries" or just "Boost".

alandefreitas commented on user-guide/modules/ROOT/pages/header-organization-compilation.adoc:
```diff
@@ -1,5 +1,5 @@
 = Header Organization and Compiled Binaries
-:navtitle: Header Organization and Compiled Binaries
+:navtitle: Headers and Binaries
```
:relieved: much better

PeterTurcan commented:
@alandefreitas I made the change - change all references to Boost C++ Library to Boost or Boost Library, added a section to the combined libraries for MySQL, and added a section on debugging.

alandefreitas commented on user-guide/modules/ROOT/pages/boost-history.adoc:
```diff
@@ -1,29 +1,29 @@
 == Introduction
 
-The Boost pass:[C++] Library is a highly influential and widely used collection of libraries, designed to extend and complement the functionality provided by the pass:[C++] Standard Library. Since its original https://www.boost.org/users/proposal.pdf[Proposal] in 1998, the Boost pass:[C++] Library has played a significant role in shaping the evolution of the pass:[C++] language and its ecosystem.
+The Boost Library is a highly influential and widely used collection of libraries, designed to extend and complement the functionality provided by the pass:[C++] Standard Library. Since its original https://www.boost.org/users/proposal.pdf[Proposal] in 1998, the Boost Library has played a significant role in shaping the evolution of the pass:[C++] language and its ecosystem.
```
> The Boost Library

It's either "Boost" or "The Boost Libraries"

PeterTurcan replied:
@alandefreitas As far as I can see this change is in my last commit - not sure which one you want me to remove?

alandefreitas replied:
Your latest commit says:

> The Boost Library is a highly influential and widely used collection of libraries, designed to extend and complement the functionality provided by the pass:[C++] Standard Library. Since its original https://www.boost.org/users/proposal.pdf[Proposal] in 1998, the Boost Library has played a significant role in shaping the evolution of the pass:[C++] language and its ecosystem.

![Uploading image.png…]()

alandefreitas replied:
It was "The Boost pass:[C++] Library" and now it is "The Boost Library"

PeterTurcan replied:
@alandefreitas  are you saying I should change this to "The Boost Libraries..." ?

alandefreitas replied:
Yes. Either "Boost" or "The Boost Libraries".

alandefreitas reviewed (changes requested):
You also need to squash the commits after this change.

PeterTurcan commented:
@alandefreitas Just Boost sounds friendlier, so I have run with that.

alandefreitas commented on user-guide/modules/ROOT/pages/visual-studio.adoc:
```diff
@@ -0,0 +1,25 @@
+* *Testing Tools*: Visual Studio's built-in testing tools can be used to test your code that uses Boost libraries. If you're using boost:test[] for unit testing, Visual Studio can recognize and run these tests within the IDE, showing the results in its *Test Explorer* window.
+
++
+In Visual Studio select *Project/Add New Item*, then select *Visual C++/Test*. You will see the boost:test[] template available to you.
```
Why are we selling the advertised benefits of VS anyway? There are so many IDEs out there and all of them have code analysis and debugging. I have to focus on selling Boost and only mention the generic benefits of some products when we have to because it's strictly related to Boost. This is only the case for the relationship between Test Explorer and Boost.Test (which other IDEs, such as CLion and VSCode, also offer).

alandefreitas commented on user-guide/modules/ROOT/pages/visual-studio.adoc:
```diff
@@ -0,0 +1,25 @@
+In Visual Studio select *Project/Add New Item*, then select *Visual C++/Test*. You will see the boost:test[] template available to you.
+
++
+image::visual-studio-boost-test.png[]
```
If anything, this should contain screenshots of test results formatted by VS rather than a screenshot of the process of getting there.

alandefreitas reviewed (changes requested):
I would say you need to squash these commits. But now there's unrelated content, so you should preferably interactively rebase squashing the first commits about libs working together and leave the content about VS in another commit. But just squashing it all is a reasonable second option. 

You can create branches for each feature so that you have PRs for each feature. The first commits are good to go, but now the review became about the last commit so the PR can't be merged. Working on `develop` all the time won't scale well.
